feat: course outline - sections and subsections list

navinkarkera opened the pull request:
This PR combines following PRs to create section and subsection list.

* See original PR: https://github.com/open-craft/frontend-app-course-authoring/pull/12 for details about section list.
* See original PR: https://github.com/open-craft/frontend-app-course-authoring/pull/20 for details about subsection list.
* See original PR: https://github.com/open-craft/frontend-app-course-authoring/pull/15 for details about section configuration modal.
* See original PR: https://github.com/open-craft/frontend-app-course-authoring/pull/14 for details about auto scroll to element.
* See original PR: https://github.com/open-craft/frontend-app-course-authoring/pull/17 for details about section drag-n-drop.

`Private-ref`: 
* https://tasks.opencraft.com/browse/BB-8154
* https://tasks.opencraft.com/browse/BB-8220
* https://tasks.opencraft.com/browse/BB-8221
* https://tasks.opencraft.com/browse/BB-8218
* https://tasks.opencraft.com/browse/BB-8219

openedx-webhooks commented:
Thanks for the pull request, @navinkarkera! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

KristinAoki commented on src/course-outline/publish-modal/PublishModal.jsx:
```diff
@@ -0,0 +1,77 @@
+        {subSections.filter(subSection => subSection.hasChanges).map((subSection) => {
+          const units = subSection.childInfo.children.filter(unit => unit.hasChanges);
+
+          return units.length ? (
```
Same question as section lists?

navinkarkera replied:
It is same, 0 is considered a falsy value.

KristinAoki commented on src/course-outline/publish-modal/PublishModal.jsx:
I will do another pass on the modal functionality when the subsections are added to the outline.

bradenmacdonald commented on src/course-outline/data/api.js:
```diff
@@ -105,3 +108,107 @@ export async function restartIndexingOnCourse(reindexLink) {
+/**
+ * Get course section
+ * @param {string} sectionId
+ * @returns {Promise<Object>}
```
Can you put the actual type here? `Object` is not very useful.

Nit/tip: Also, it's not your code but there are some major problems with the JSDoc comments in the other functions above. if you put `// @ts-check` on the first line of this function, you can easily see all the errors using an IDE like VS Code.

e.g.

![Screenshot 2023-12-08 at 3 00 45 PM](https://github.com/openedx/frontend-app-course-authoring/assets/945577/f63626ea-60fb-4e00-b7fb-5eb0eef5e504)

Should be:

```ts
/**
 * Get course best practices.
 * @param {{courseId: string, excludeGraded: boolean, all: boolean}} options
 * @returns {Promise<{isSelfPaced: boolean, sections: any }>}
 */
export async function getCourseBestPractices({
  courseId,
  excludeGraded,
  all,
}) {
```

You don't have to change that now, but I do strongly recommend putting in the `// @ts-check` so that you're not adding more errors.

navinkarkera replied:
@bradenmacdonald Thanks! Actually, even this is not my code :sweat_smile:. But yes, I should have checked these docs. Will try to fix them, atleast in course-outline related files.

bradenmacdonald replied:
haha I see. Well thanks for making those fixes anyways :)
I'm not planning to do any further review here, since @KristinAoki has it covered. But let me know if you do have specific questions I can help with.

bradenmacdonald commented on src/course-outline/constants.js:
```diff
@@ -1,9 +1,25 @@
+export const SECTION_BADGE_STATUTES = {
+  live: 'live',
+  publishedNotLive: 'published_not_live',
+  staffOnly: 'staff_only',
+  draft: 'draft',
+};
```
```suggestion
export const SECTION_BADGE_STATUTES = /** @type {const} */ ({
  live: 'live',
  publishedNotLive: 'published_not_live',
  staffOnly: 'staff_only',
  draft: 'draft',
});
```
Do this and you'll get the correct types :)

![Screenshot 2023-12-08 at 3 10 58 PM](https://github.com/openedx/frontend-app-course-authoring/assets/945577/e5d3ee5b-0306-41ec-b9ab-53bebedbf54a)

bradenmacdonald commented on src/course-outline/utils.jsx:
```diff
@@ -0,0 +1,116 @@
+ * @param {bool} visibleToStaffOnly - value from section info
+ * @param {string} visibilityState - value from section info
+ * @param {bool} staffOnlyMessage - value from section info
+ * @returns {typeof SECTION_BADGE_STATUTES}
```
```suggestion
 * @returns {SECTION_BADGE_STATUTES[keyof SECTION_BADGE_STATUTES]}
```
What you have is wrong - it's returning an `Object` when this is a `string`. Use this ^ along with the other fix I suggested and you actually get the right types:

![Screenshot 2023-12-08 at 3 12 31 PM](https://github.com/openedx/frontend-app-course-authoring/assets/945577/98976b40-a282-466a-9890-c2baaab96647)

Throw in `// @ts-check` at the top of `utils.jsx` and `SectionCard.tsx` and you can see these issues more easily in VS Code.

navinkarkera commented on src/course-outline/highlights-modal/HighlightsModal.jsx:
```diff
@@ -0,0 +1,94 @@
+                })}
+              </p>
+              {Object.entries(initialFormValues).map(([key], index) => (
+                <FormikControl
```
@KristinAoki It renders an input text box as `FormikControl` uses `Form.Control`.

KristinAoki replied:
Sorry, I meant `TextArea`. It would be nice to see a multiline view.

navinkarkera commented on src/course-outline/CourseOutline.jsx:
```diff
@@ -97,6 +133,37 @@ const CourseOutline = ({ courseId }) => {
                       openEnableHighlightsModal={openEnableHighlightsModal}
                     />
+                    <div className="pt-4">
+                      {sectionsList.length ? (
```
It is same, 0 is considered a falsy value.

navinkarkera commented:
@KristinAoki @bradenmacdonald I have addressed your comments

@KristinAoki Also combined few PRs (linked in description) for you to review and merge them in a single go as discussed in the meeting.

KristinAoki commented on src/course-outline/utils.jsx:
```diff
@@ -110,7 +110,11 @@ const getHighlightsFormValues = (currentHighlights) => {
 };
 
 const scrollToElement = (ref) => {
```
Had a weird bug appear when editing this. Repro steps
1. Add a new section
2. Should scroll to the new section
3. Scroll up to another section and edit its visibility
Expected
- Once everything finishes saving, page should remain on edited section
Actual
- scrolls to the recently added section

This also happens when editing in the basic tab

navinkarkera replied:
@KristinAoki This should be fixed now. We will only scroll to element when required.

KristinAoki commented on src/course-outline/utils.jsx:
```diff
@@ -110,7 +110,11 @@ const getHighlightsFormValues = (currentHighlights) => {
 
 const scrollToElement = (ref) => {
-  ref.current?.scrollIntoView({ behavior: 'smooth' });
+  ref.current?.scrollIntoView({
```
Do we want this to scroll even if the element is already in view? When I have a section in the center of my screen and I add a new subsection, the page scrolls so that the new subsection is at the bottom of my screen and the add button is no longer visible.

navinkarkera replied:
@KristinAoki Updated to only scroll of element is not visible.

navinkarkera commented:
@KristinAoki Addressed your comments about scrolling. We have also merged section drag-n-drop PR to this one. Please take a look. I'll fix failed tests tomorrow morning.

KristinAoki reviewed (commented):
Regarding the drag-and-drop, [there is already a component](https://github.com/openedx/frontend-lib-content-components/tree/main/src/editors/sharedComponents/DraggableList) in `frontend-lib-content-components` that does all of the configuration for the dnd library. It is being used in `custom-pages` for the page reordering.

navinkarkera commented:
> Regarding the drag-and-drop, [there is already a component](https://github.com/openedx/frontend-lib-content-components/tree/main/src/editors/sharedComponents/DraggableList) in `frontend-lib-content-components` that does all of the configuration for the dnd library. It is being used in `custom-pages` for the page reordering.

@KristinAoki Thanks, did not know this. Updated to use these components similar to custom-pages. The only difference now is that we have a drag handle instead of allowing the user to drag the section from any point. Even if it is a bit different than the design, it is better as it will reduce complexity while making subsections and units movable.

![image](https://github.com/openedx/frontend-app-course-authoring/assets/10894099/85c0ed2c-73a7-480b-bc96-0c2bdd91a0f0)

KristinAoki reviewed (commented):
Drag and drop for sections looks great!

navinkarkera commented:
@KristinAoki Thanks! I think this is ready to be merged if it looks good to you. Just let me know before merging, I'll squash commits as per the features and we can rebase and merge it.

KristinAoki reviewed (commented):
Once code coverage is passing, I will approve the PR

navinkarkera commented:
@KristinAoki This is ready to be merged now. Squashed commits as per feature and authors as much as possible, please use rebase and merge option while merging.

openedx-webhooks commented:
@navinkarkera 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-course-authoring&entry.867055334=https://github.com/openedx/frontend-app-course-authoring/pull/733&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2023-12-07+07:33&entry.1917517419=2023-12-20+15:20&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
